staff/acct/sorry/sorry

```diff
@@ -1,156 +1,183 @@
+    if attributes is None:
+        raise ValueError('User not found.')
+
+    rootstaffer = real_user()
+    kerberos_credentials(rootstaffer)
+
+    user_dir = attributes['homeDirectory']
+
+    # Workaround for ldap3 1.x.x machines
+    if isinstance(user_dir, list):
+        user_dir = user_dir[0]
+
+    print('Copying sorry file and making .oldshell file')
+
+    if not (os.path.isfile(user_dir + '/.oldshell')):
+        oldshell = attributes['loginShell']
+        # Workaround for ldap3 1.x.x machines
+        if isinstance(oldshell, list):
+            oldshell = user_dir[0]
+
+        oldshell_file = open(user_dir + '/.oldshell', 'w')
+        oldshell_file.write(oldshell)
+        oldshell_file.close()
+
+    shutil.copy2(sorry_file, user_dir + '/.sorry')
+    os.chmod(user_dir + '/.sorry', 0o400)
+    shutil.chown(user_dir + '/.sorry', user=username, group='ocf')
+
+    print("Changing user's shell to sorry shell")
+
+    ocflib.infra.ldap.modify_ldap_entry(ocflib.account.utils.dn_for_username(username),
+                                        {'loginShell': sorry_shell, 'gidNumber': grp.getgrnam('sorry').gr_gid})
+
+    http_dir = '/services/http/users/{0}/{1}'.format(username[0], username)
+    print('Changing permissions on httpdir to 000')
+    if (os.path.isdir(http_dir)):
+        os.chmod(http_dir, 0o000)
+
+    print('Changing permissions on home directory to 500')
+    os.chmod(user_dir, 0o500)
+
+    # WTF is final system check for?
+    # log(rootstaffer, username)
+    send_log_mail(rootstaffer, username, open(reason_path, 'r').read())
+
+    user_mail = get_user_real_mail(username)
+    if user_mail:
+        send_sorry_mail(user_mail, username, open(reason_path, 'r').read())
+    kerberos_destroy()
+
+# Returns the real user who is running this script
+
+
+def real_user():
+    if 'SUDO_USER' in os.environ and os.environ['SUDO_USER'] != 'root':
+        return os.environ['SUDO_USER']
+    else:
+        print("The sorry.log is much more useful when it logs who you are rather than simply 'root'.")
+        username = input('Please enter your username: ')
+        return username
+
+
+def kerberos_credentials(staff_name):
+    if ('SORRY_KRB5CCNAME' in os.environ):
+        print('SORRY_KRB5CCNAME set in environment. Assuming this file contains current admin credentials.')
+        os.environ['KRB5CCNAME'] = os.environ['SORRY_KRB5CCNAME']
+    else:
+        print('Your are {0}.'.format(staff_name))
+        if (subprocess.call([shutil.which('kinit'), '{0}/admin'.format(staff_name)]) != 0):
+            raise OSError('Kinit failed, bailing out!')
+        else:
+            print('Kinit complete.')
+
+
+def kerberos_destroy():
+    return subprocess.call(['kdestroy'])
+
+
+def get_user_real_mail(username):
+    email = str(subprocess.run([shutil.which('ldapsearch'),
+                                'uid={0}'.format(username), 'mail'], stdout=subprocess.PIPE).stdout)
```

Comment:
This runs fine, what will change if we use .decode()?

Reply:
Casting bytes to string like this gives you this terrible thing that looks like
```
"b'# extended LDIF\\n#\\n# LDAPv3\\n# base <dc=ocf,dc=berkeley,dc=edu> (default) with scope subtree\\n# filter: uid=cooperc\\n# requesting: mail \\n#\\n\\n# cooperc, People, OCF.Berkeley.EDU\\ndn: uid=cooperc,ou=People,dc=OCF,dc=Berkeley,dc=EDU\\nmail: cooperc@g.ocf.berkeley.edu\\n\\n# search result\\nsearch: 5\\nresult: 0 Success\\n\\n# numResponses: 2\\n# numEntries: 1\\n'"
```
Using `.decode()`, which is the best practice way to convert bytes to str will give
```
'# extended LDIF\n#\n# LDAPv3\n# base <dc=ocf,dc=berkeley,dc=edu> (default) with scope subtree\n# filter: uid=cooperc\n# requesting: mail \n#\n\n# cooperc, People, OCF.Berkeley.EDU\ndn: uid=cooperc,ou=People,dc=OCF,dc=Berkeley,dc=EDU\nmail: cooperc@g.ocf.berkeley.edu\n\n# search result\nsearch: 5\nresult: 0 Success\n\n# numResponses: 2\n# numEntries: 1\n'
```

Basically `str()` will literally give you the string representation of the bytes, rather than actually converting it.

Reply:
I'll use jvperrin's function directly...